### ISSUE TITLE
Trace/Span ids base64 encoded when used in proto-JSON representation

### ISSUE THREAD

**Issue by bogdandrutu** [open-telemetry/opentelemetry-proto#129]
As pointed in https://github.com/open-telemetry/opentelemetry-proto/pull/110 currently trace_id and span_id are bytes which requires base64 encoded in a proto-JSON representation. This makes them look in JSON very different from the hex representation used in Trace-Context. Here are some options:
1. Change trace/span id to string hex representation;
2. Define a different json schema for otlp HTTP;
3. Try to push to protobuf support custom encoding for bytes fields;

**Comment by Oberon00** [comment]
In theory, I would love for W3C to use base64 instead of hex. After all it's shorter and at some cloud vendor setups, I've heard, you are paying per byte you send/receive (e.g. if the HTTP request is ultimately serialized to a messaging system).

**Comment by SergeyKanzhelev** [comment]
> I would love for W3C to use base64 instead of hex.

@Oberon00 I don't think changing W3C is feasible at this stage. For history - base64 vs hex was heavily debated. Unfortunately this debate hasn't made it to the [rationale](https://github.com/w3c/trace-context/blob/master/http_header_format_rationale.md), but as overview - we strived for human-readability and consistency with prior art.

@bogdandrutu what's the pros/cons of each option? 

> Change trace/span id to string hex representation;

This is 4x size and CPU on serialization/deserialization, correct? Also potential abuse with other characters. Any other issues?

> Define a different json schema for otlp HTTP;

Have you looked at the whole schema? Any other issues with JSON?

> Try to push to protobuf support custom encoding for bytes fields;

I have no context on this. Can you comment on pros/cons?

**Comment by tigrannajaryan** [comment]
>Change trace/span id to string hex representation;

I don't think we should do this and slow-down / increase wire size for all other languages.

**Comment by SergeyKanzhelev** [comment]
@tigrannajaryan to assess the second option - can you generate an example of a JSON from proto? So we can review for other potential issues?

Can you please also comment on option 3 viability?

**Comment by tigrannajaryan** [comment]
@SergeyKanzhelev unfortunately I don't know how 2 and 3 done, I have no experience with that. It is best that someone who knows what they are doing works on this.

**Comment by tensorchen** [comment]
@SergeyKanzhelev @tigrannajaryan  This is an  example of a JSON from proto.

```json
{
        "resource_spans": [
                {
                        "instrumentation_library_spans": [
                                {
                                        "spans": [
                                                {
                                                        "trace_id": "X4qEKE3/fr2Vi6ZQOWTCyg==",
                                                        "span_id": "whYPphcCSbc=",
                                                        "parent_span_id": "UrOJGGtlMGU=",
                                                        "name": "Sample-8",
                                                        "kind": 1,
                                                        "start_time_unix_nano": 1588324858573252000,
                                                        "end_time_unix_nano": 1588324858579272663,
                                                        "status": {}
                                                }
                                        ]
                                }
                        ]
                }
        ]
}

```

In spec https://github.com/open-telemetry/oteps/blob/master/text/0099-otlp-http.md

Does it only support application/x-protobuf,  not application/json?

**Comment by tigrannajaryan** [comment]
@tensorchen Correct, OTLP/HTTP currently defines only "application/x-protobuf". If there is a desire to send JSON payload that needs to be explored and added as an extension to the protocol. I am not sure what exactly are the implications of that so that will require some investigation.

**Comment by nilebox** [comment]
@bogdandrutu @tigrannajaryan What's the current state of HTTP/JSON support in OTLP?
Can we start working on supporting it on the client (SDK) side, or should we wait until this issue is resolved?

**Comment by bogdandrutu** [comment]
Would not start to support yet HTTP/JSON as another protocol, we currently have gRPC/Protobuf, gRPC over HTTP (with the limitation of the trace/span id), and HTTP/Protobuf.

**Comment by nilebox** [comment]
Raised https://github.com/open-telemetry/opentelemetry-collector/issues/1177 to disable support for JSON encoding in the Collector until we decide to officially support it.

Shall we also remove the Swagger code generation in https://github.com/open-telemetry/opentelemetry-proto/blob/e43e1abc40428a6ee98e3bfd79bec1dfa2ed18cd/makefile#L50-L55 ?

**Comment by nilebox** [comment]
Since Collector is now using `gogo/protobuf`, we may configure protos to generate trace/span IDs with a custom type via https://github.com/gogo/protobuf/blob/master/custom_types.md

It shouldn't lead to any issues with binary wire compatibility, but the solution is specific to Go, so SDKs for other languages won't benefit from this.

**Comment by tigrannajaryan** [comment]
I am going to assign this to myself to have a look. My preliminary feeling is that since W3C Trace context uses hex then we should aim to do the same (as @SergeyKanzhelev rightfully points out it is too late to change W3C Trace context). I do not know how feasible it is to change Protobuf's to use hex encoding instead of base64, I will look into this.

**Comment by tigrannajaryan** [comment]
Traceid/spanid are also part of the Trace protocol which we declared stable. I think it is OK that we make this change (should we decide that it is desirable and doable) and we are not actually breaking our promise for Trace protocol stability which is only about binary Protobuf encoding and not JSON encoding which was introduced very recently (https://github.com/open-telemetry/oteps/pull/122) and where this issue was not debated at all.
